### PR TITLE
Several improvements to logging and debugging features.

### PR DESCRIPTION
**Pull request by joshw**
* more detail, including responses, on zenopenstack diagnostic pages
  * added device links from openstack device to these diagnostics
  * improved error logging in zenopenstack
  * added twisted manhole to zenopenstack

### CODE REVIEW

**Comment by joshw** [comment]
(bump)

File: ZenPacks/zenoss/OpenStackInfrastructure/Endpoint.py [hunk 1]
```diff
@@ -164,8 +163,39 @@ def public_keystone_apiendpoint(self):
         except AttributeError:
             return None
 
+    def zenopenstack_url(self, https=False):
+        try:
+            from Products.ZenUtils.controlplane.application import getConnectionSettings
+            from Products.ZenUtils.controlplane import ControlPlaneClient
+        except ImportError:
+            return
```
Comment (pcarinhas) [inline ZenPacks/zenoss/OpenStackInfrastructure/Endpoint.py]: Include a log here or where used for debugging..

File: ZenPacks/zenoss/OpenStackInfrastructure/zenopenstack.py [hunk 2]
```diff
@@ -300,18 +300,52 @@ def render_GET(self, request):
...
 
             ip = request.postpath[1]
             uri = "/".join(request.postpath[2:])
             body = "<html><body>"
             records = self.site.request_buffer.get_requests((ip, '/' + uri))
-            body += "<b>Last %d requests from /%s to %s</b> (of %d max)<p>" % (
+            body += "<b>Most recent %d requests from /%s to %s</b> (of %d max)<p>" % (
                 len(records),
                 ip,
                 uri,
```
Comment (pcarinhas) [inline ZenPacks/zenoss/OpenStackInfrastructure/zenopenstack.py]: Its getting a bit hairy down here with the presence of preformatted body string. Not sure how to clean that up.

File: ZenPacks/zenoss/OpenStackInfrastructure/zenopenstack.py [hunk 3]
```diff
@@ -300,18 +300,52 @@ def render_GET(self, request):
...
+                    body += "<hr>%s (%s ago)" % (
+                        timestamp.isoformat(),
+                        (datetime.datetime.now() - timestamp)
+                    )
+                    body += "<pre>" + cgi.escape(pformat(objmap)) + "</pre>"
+
+                body += "</body></html>"
+                return body
+
+
```
[the remark this answers is not in the document]
Comment (joshw) [reply ZenPacks/zenoss/OpenStackInfrastructure/zenopenstack.py]: fixed

File: ZenPacks/zenoss/OpenStackInfrastructure/Endpoint.py [hunk 4]
```diff
@@ -164,8 +163,39 @@ def public_keystone_apiendpoint(self):
...
+            endpoint_name = "proxy-zenopenstack_https"
+        else:
+            protocol = "http"
+            endpoint_name = "proxy-zenopenstack_http"
+
+        collector = self.perfServer().id
+        client = ControlPlaneClient(**getConnectionSettings())
+        for svc_id in [x.id for x in client.queryServices() if x.name == 'proxy-zenopenstack']:
+            svc = client.getService(svc_id)
+            for endpoint in filter(lambda s: s['Name'] == endpoint_name, svc.getRawData()['Endpoints']):
```
[the remark this answers is not in the document]
Comment (joshw) [reply ZenPacks/zenoss/OpenStackInfrastructure/Endpoint.py]: As far as I am aware, there is no need in this case.  (The serialized object being returned will always have an 'Endpoints')

File: ZenPacks/zenoss/OpenStackInfrastructure/Endpoint.py [hunk 5]
```diff
@@ -164,8 +163,39 @@ def public_keystone_apiendpoint(self):
...
+                try:
+                    if client.getService(svc.parentId).name == collector:
+                        return "{protocol}://{ip}:{port}".format(
+                            protocol=protocol,
+                            ip=endpoint['AddressAssignment']['IPAddr'],
+                            port=endpoint['AddressAssignment']['Port'])
```
[the remark this answers is not in the document]
Comment (joshw) [reply ZenPacks/zenoss/OpenStackInfrastructure/Endpoint.py]: I don't think there's a need.   CC is pretty strict on what it returns.

File: ZenPacks/zenoss/OpenStackInfrastructure/Endpoint.py (hunk 1 above)
Comment (joshw) [reply ZenPacks/zenoss/OpenStackInfrastructure/Endpoint.py]: what isn't logged that should be?

File: ZenPacks/zenoss/OpenStackInfrastructure/zenopenstack.py (hunk 2 above)
Comment (joshw) [reply ZenPacks/zenoss/OpenStackInfrastructure/zenopenstack.py]: Yeah, in general the /health stuff is gnarly.  I'd like to rework it eventially to be more of a json API and all this html stuff moved into a static html file with javascript, but that is currently not something I have time to do.

**Comment by joshw** [comment]
I'd like QA to have this tomorrow AM, so i'm going to merge now and we can revisit any concerns in a followup PR

**Pull request merged**